### ISSUE TITLE
Time Disputes

### ISSUE THREAD

**Issue by eskimor** [paritytech/polkadot-sdk#742]
Latest design [here](https://github.com/paritytech/polkadot-sdk/issues/742#issuecomment-1691884835).

# Basic Idea/Design Considerations

## New objectives:

1. It is not a security threat for approval checkers to not vote for a very long
   time, because it is taking long - approval voting is secure regardless.
2. We do need some timeout for liveness only, without a dispute we would never
   finalize.
3. Backers are special, they are the only validators that are forced to commit
   to a <2s execution.
4. Approval checkers will send timing information, if it took them significantly longer than
   the backing timeout then this is put on chain.
5. We charge backers via era points for that extra time. The charging will be
   exponential and reaches the cost of raising a concluding valid dispute. Thus
   we can actually have _very_ long approval voting timeouts.
6. Voting invalid validators will not get charged for wasting everybody's time,
   if that bill has already been paid by the backers via 5. (Getting honest
   validators slashed is thus no longer possible.) Basically validators voting
   invalid only need to get charged whatever the backers were not charged yet.
7. On disputes concluding invalid, backers always get slashed 100%, voting valid
   approval checkers (and dispute participants) only if there is no (small
   enough) time report of approval checkers. This way there is an incentive to
   actually validate (and not just always vote valid  -which would be the case if we always only slashed backers).

### Questions?

1. Would it be possible for malicious approval checkers to get honest backers
   charged? Assuming enough honest approval checkers, this should not be
   possible.

### Threats/Complications:

Getting those approval votes (time) on chain:

1. How many? Only tranch 0? Is that secure, what if validators are DoSed?
2. Cherry picking?
3. If we allow for a very long approval voting timeout, we might run into issues with state pruning. Let's assume an attacker tries to get a bad candidate in. It makes it so, that it takes a very long time on honest checkers. Now if he sees that approval checkers are not his guys, the attacker reveals a better fork (not including that candidate) - that chain gets finalized. Now by the time the validation for the fork timed out, the chain's state has already been discarded (too far behind the finalized block height), which might prevent a dispute from succeeding. In fact nodes might not even participate at all, as they have not "seen" the candidate backed/included.  Both should be fixable though, by not clearing up too eagerly on finalization.

Tl;DR: By charging backers for excess time proportional to the amount, we can afford very long approval checking timeouts. This way natural fluctuations in load/performance should no longer realistically cause a dispute and even if that timeout is triggered by malicious backers, they would end up paying the bill. We would slash backers on concluding invalid disputes, but also other nodes voting valid if the timing reported by approval checkers is low enough, that it can be deemed unlikely that the nodes who voted invalid did so because of the timeout. This way approval checkers keep having an incentive to actually do the validation and not just always vote valid (which would be risk free, if we only ever slashed backers).

**Comment by Overkillus** [comment]
This will still relevant is very low prio due to the new disabling strategy mitigating a lot of the risks (combined with gov refunds in case of attacks). PVM should also heavily reduce the risk further.

**Comment by burdges** [comment]
We've no work planned because polkavm could hopefully make this irrelevent
